python/kvikio/_lib/libnvcomp.pyx

```diff
@@ -57,305 +63,184 @@ class pyNvcompType_t(Enum):
-        temp_arr,
-        temp_bytes,
-        out_arr,
-        out_bytes,
-        uintptr_t stream=0
-    ):
-        cdef uintptr_t in_ptr=Array(in_arr).ptr
-        cdef uintptr_t temp_ptr=Array(temp_arr).ptr
-        cdef uintptr_t out_ptr=Array(out_arr).ptr
-        cdef uintptr_t out_bytes_ptr=Array(out_bytes).ptr
-        self.c.compress_async(
-            <void*>in_ptr,
-            <size_t>in_bytes,
-            <void*>temp_ptr,
-            <size_t>temp_bytes,
-            <void*>out_ptr,
-            <size_t*>out_bytes_ptr,
-            <cudaStream_t>stream)
-
-cdef class _CascadedDecompressor:
-    cdef __CascadedDecompressor* d
-
-    def __cinit__(self):
-        self.d = new __CascadedDecompressor()
-
-    def __dealloc__(self):
-        del self.d
-
-    cpdef configure(self, in_arr, in_bytes, temp_bytes, out_bytes, uintptr_t stream=0):
-        cdef uintptr_t in_ptr = Array(in_arr).ptr
-        cdef uintptr_t temp_bytes_ptr = Array(temp_bytes).ptr
-        cdef uintptr_t out_bytes_ptr = Array(out_bytes).ptr
-        self.d.configure(
-            <void*>in_ptr,
-            <size_t>in_bytes,
-            <size_t*>temp_bytes_ptr,
-            <size_t*>out_bytes_ptr,
-            <cudaStream_t>stream)
+        comp_buffer
+    ) -> dict:
+        cdef shared_ptr[DecompressionConfig] partial = make_shared[
+            DecompressionConfig](self._impl.configure_decompression(
+                <uint8_t*><uintptr_t>comp_buffer.data.ptr
+            )
+        )
+        self._decompression_config = make_shared[DecompressionConfig](
+            (move(partial.get()[0]))
+        )
+        return {
+            "decomp_data_size": self._decompression_config.get()[0].decomp_data_size,
+            "num_chunks": self._decompression_config.get()[0].num_chunks
+        }
 
-    def decompress_async(
+    def decompress(
         self,
-        in_arr,
-        in_bytes,
-        temp_arr,
-        temp_bytes,
-        out_arr,
-        out_bytes,
-        uintptr_t stream=0
+        decomp_buffer,
+        comp_buffer,
     ):
-        cdef uintptr_t in_ptr = Array(in_arr).ptr
-        cdef uintptr_t temp_ptr = Array(temp_arr).ptr
-        cdef uintptr_t out_ptr = Array(out_arr).ptr
-        self.d.decompress_async(
-            <void*>in_ptr,
-            <size_t>in_bytes,
-            <void*>temp_ptr,
-            <size_t>temp_bytes,
-            <void*>out_ptr,
-            <size_t>out_bytes,
-            <cudaStream_t>stream)
-
+        self._impl.decompress(
+            <uint8_t*><uintptr_t>decomp_buffer.data.ptr,
+            <const uint8_t*><uintptr_t>comp_buffer.data.ptr,
+            <DecompressionConfig&>self._decompression_config.get()[0]
+        )
 
-# LZ4 Compressor / Decompressor
-cdef class _LZ4Compressor:
-    cdef __LZ4Compressor* c
+    def set_scratch_buffer(self, new_scratch_buffer):
+        return self._impl.set_scratch_buffer(
+            <uint8_t*><uintptr_t>new_scratch_buffer.data.ptr
+        )
```

Comment:
Would suggest hanging onto this to ensure it stays alive.

Also do we need to do anything in the manager to release the old buffer?

```suggestion
    def set_scratch_buffer(self, Array new_scratch_buffer):
        self._scratch_buffer = new_scratch_buffer
        return self._impl.set_scratch_buffer(
            <uint8_t*>self._scratch_buffer.ptr
        )
```

Reply:
@madsbk has this same conversation with @eschmidt-nvidia earlier. The C++ library has no ownership over the scratch buffer and the caller is free to do with it as they please. As above it is owned by the caller of `nvcomp.py`.